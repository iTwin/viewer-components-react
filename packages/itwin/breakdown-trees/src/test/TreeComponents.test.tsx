/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import * as React from "react";
import { Provider } from "react-redux";
import { mount, shallow } from "enzyme";
import type { IModelConnection, ScreenViewport, SelectionSet, ViewState } from "@itwin/core-frontend";
import { IModelApp, NoRenderApp } from "@itwin/core-frontend";
import { BeEvent } from "@itwin/core-bentley";
import Rules from "../assets/SpatialBreakdown.json";
import type { TreeWithRuleSetController } from "./Utils";
import { setupDataProvider, TestUtils } from "./Utils";
import * as moq from "typemoq";
import type { RegisteredRuleset, Ruleset } from "@itwin/presentation-common";
import type { IModelHierarchyChangeEventArgs, PresentationManager, RulesetManager, RulesetVariablesManager } from "@itwin/presentation-frontend";
import { Presentation } from "@itwin/presentation-frontend";
import type { TreeActions, TreeModel, TreeModelNode } from "@itwin/components-react";
import { PropertyValueRendererManager } from "@itwin/components-react";
import sinon from "sinon";
import { TreeNodeFunctionIconInfoMapper } from "../Views/FunctionalityProviders/TreeNodeFunctionIconMapper";
import { ControlledTreeWrapper } from "../Views/TreeWithRuleset";
import { TreeNodeWrapper } from "../Views/NodeRenderers/FunctionalTreeNodeRenderer";
import { ClassificationsTree } from "../Views/ClassificationsTree";
import { ComponentIndex } from "../Views/ComponentIndex";
import { SpatialContainmentTree } from "../Views/SpatialContainmentTree";
import { expect } from "chai";

describe("TreeComponent tests.", () => {

  const connection = moq.Mock.ofType<IModelConnection>();
  const controller = moq.Mock.ofType<TreeWithRuleSetController>();
  before(async () => {
    await NoRenderApp.startup();
    try {
      await Presentation.initialize();
    } catch (error) {
    }

    const selectionSet = moq.Mock.ofType<SelectionSet>();
    selectionSet.setup((x) => x.elements).returns(() => new Set([]));
    connection.setup((x) => x.selectionSet).returns(() => selectionSet.object);
    await TestUtils.initializeUiFramework(connection.object);
    await IModelApp.localization.registerNamespace("BreakdownTrees");
    controller.setup((x) => x.createDataProvider()).returns(() => setupDataProvider(connection.object));

    onHierarchyUpdateEvent = new BeEvent();
    presentationManagerMock.setup((x) => x.onIModelHierarchyChanged).returns(() => onHierarchyUpdateEvent);
    presentationManagerMock.setup((x) => x.rulesets()).returns(() => rulesetManagerMock.object);
    presentationManagerMock.setup((x) => x.vars(moq.It.isAny())).returns(() => rulesetVariablesManagerMock.object);
    Presentation.setPresentationManager(presentationManagerMock.object);

    const selectedViewMock = moq.Mock.ofType<ScreenViewport>();
    const viewStateMock = moq.Mock.ofType<ViewState>();
    viewStateMock.setup((x) => x.isSpatialView()).returns(() => false);
    selectedViewMock.setup((x) => x.view).returns(() => viewStateMock.object);
    await IModelApp.viewManager.setSelectedView(selectedViewMock.object);
  });

  const rulesetManagerMock = moq.Mock.ofType<RulesetManager>();
  const registeredRuleSetMock = moq.Mock.ofType<RegisteredRuleset>();
  const rulesetVariablesManagerMock = moq.Mock.ofType<RulesetVariablesManager>();
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();

  let onHierarchyUpdateEvent: BeEvent<(args: IModelHierarchyChangeEventArgs) => void>;
  beforeEach(async () => {
    rulesetVariablesManagerMock.reset();
    rulesetVariablesManagerMock.setup((x) => x.onVariableChanged).returns(() => new BeEvent());
  });

  after(async () => {
    (global as any).sessionStorage = undefined;
    TestUtils.terminateUiFramework();
    try {
      Presentation.terminate();
      await IModelApp.shutdown();
    } catch (error) { }
  });

  it("TreeWithRuleset renders correctly", () => {
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ControlledTreeWrapper
          dataProvider={dataProvider}
          setIsDisplayGuids={(_display: boolean) => { }}
          iModel={connection.object}
          displayGuids={true}
          loadedRuleset={Rules as Ruleset} treeNodeIconMapper={functionIconMapper}
          treeName="TreeWithRulesetTest" optionItems={[]} searchTools={true} enableVisibility={false}
        />
      </Provider>
    );
    shallow(wrapper).should.matchSnapshot();
  });

  it("TreeWithRuleset renders correctly with visibility", () => {
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ControlledTreeWrapper
          dataProvider={dataProvider}
          setIsDisplayGuids={(_display: boolean) => { }}
          iModel={connection.object}
          displayGuids={true}
          loadedRuleset={Rules as Ruleset} treeNodeIconMapper={functionIconMapper}
          treeName="TreeWithRulesetTest" optionItems={[]} searchTools={true} enableVisibility={true}
        />
      </Provider>
    );
    shallow(wrapper).should.matchSnapshot();
  });

  it("TreeNodeWrapper renders correctly", () => {
    const node = moq.Mock.ofType<TreeModelNode>();
    const treeModel = moq.Mock.ofType<TreeModel>();
    const treeAction = moq.Mock.ofType<TreeActions>();
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const reactNode = moq.Mock.ofType<React.ReactNode>();
    const propertyValueRenderStub = sinon.stub(PropertyValueRendererManager.defaultManager, "render").returns(reactNode.object);

    const wrapper = (
      <TreeNodeWrapper node={node.object} treeActions={treeAction.object} itemsMapper={functionIconMapper} visibilityHandler={undefined} treeModel={treeModel.object}
        selectedTreenodeCount={0} />
    );
    mount(wrapper).should.matchSnapshot();
    reactNode.reset();
    propertyValueRenderStub.restore();
  });
  it("ClassificationsTree renders correctly with children", () => {
    rulesetManagerMock.setup(async (x) => x.add(moq.It.isAny())).returns(async () => registeredRuleSetMock.object);
    const wrapper = (
      <ClassificationsTree iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });

  it("ComponentIndex renders correctly with children", () => {
    const wrapper = (
      <ComponentIndex iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });
  it("SpatialContainmentTree renders correctly with children", () => {
    const wrapper = (
      <SpatialContainmentTree iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }}
        groupByType={false} groupByDiscipline={false} setGroupByType={(_groupByType: boolean) => { }} setGroupByDiscipline={(_groupByDiscipline: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });

  it("SpatialContainmentTree by Type renders correctly with children", () => {
    const wrapper = (
      <SpatialContainmentTree iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }}
        groupByType={true} groupByDiscipline={false} setGroupByType={(_groupByType: boolean) => { }} setGroupByDiscipline={(_groupByDiscipline: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });

  it("SpatialContainmentTree by Discipline renders correctly with children", () => {
    const wrapper = (
      <SpatialContainmentTree iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }}
        groupByType={false} groupByDiscipline={true} setGroupByType={(_groupByType: boolean) => { }} setGroupByDiscipline={(_groupByDiscipline: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });

  it("SpatialContainmentTree by Type and Discipline renders correctly with children", () => {
    const wrapper = (
      <SpatialContainmentTree iModel={connection.object} displayGuids={false} setIsDisplayGuids={(_displayGuids: boolean) => { }}
        groupByType={true} groupByDiscipline={true} setGroupByType={(_groupByType: boolean) => { }} setGroupByDiscipline={(_groupByDiscipline: boolean) => { }} />
    );
    expect(wrapper).matchSnapshot();
  });

});
