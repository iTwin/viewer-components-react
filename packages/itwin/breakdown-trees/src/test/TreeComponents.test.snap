// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeComponent tests. ClassificationsTree renders correctly with children 1`] = `
<ClassificationsTree
  displayGuids={false}
  iModel={null}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. ComponentIndex renders correctly with children 1`] = `
<ComponentIndex
  displayGuids={false}
  iModel={null}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. SpatialContainmentTree by Discipline renders correctly with children 1`] = `
<SpatialContainmentTree
  displayGuids={false}
  groupByDiscipline={true}
  groupByType={false}
  iModel={null}
  setGroupByDiscipline={[Function]}
  setGroupByType={[Function]}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. SpatialContainmentTree by Type and Discipline renders correctly with children 1`] = `
<SpatialContainmentTree
  displayGuids={false}
  groupByDiscipline={true}
  groupByType={true}
  iModel={null}
  setGroupByDiscipline={[Function]}
  setGroupByType={[Function]}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. SpatialContainmentTree by Type renders correctly with children 1`] = `
<SpatialContainmentTree
  displayGuids={false}
  groupByDiscipline={false}
  groupByType={true}
  iModel={null}
  setGroupByDiscipline={[Function]}
  setGroupByType={[Function]}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. SpatialContainmentTree renders correctly with children 1`] = `
<SpatialContainmentTree
  displayGuids={false}
  groupByDiscipline={false}
  groupByType={false}
  iModel={null}
  setGroupByDiscipline={[Function]}
  setGroupByType={[Function]}
  setIsDisplayGuids={[Function]}
/>
`;

exports[`TreeComponent tests. TreeNodeWrapper renders correctly 1`] = `
<TreeNodeWrapper
  highlightingProps={
    Object {
      "searchText": "tree",
    }
  }
  itemsMapper={
    TreeNodeFunctionIconInfoMapper {
      "_TreeNodeClassFunctionIconInfoMap": Map {},
      "_functionIconInfos": Array [],
      "_globalFunctionIconInfos": Array [],
      "_groupNodeFunctionIconInfos": Array [],
      "_multipleNodesFunctionIconInfos": Array [],
      "_treeDataProvider": Object {
        "dispose": [Function],
        "getFilteredNodePaths": [Function],
        "getNodeKey": [Function],
        "getNodes": [Function],
        "getNodesCount": [Function],
        "imodel": null,
        "rulesetId": "breakdown-trees-react/RulesList",
      },
    }
  }
  node={null}
  selectedTreenodeCount={0}
  treeActions={null}
  treeModel={null}
>
  <div
    data-testid="nodeLoaded"
    onContextMenu={[Function]}
  >
    <div>
      <TreeNode
        isExpanded={[Function]}
        isLeaf={false}
        isLoading={[Function]}
        isSelected={[Function]}
        key="TreeNodeWrapper"
        label={
          <div
            className={undefined}
          >
            null
          </div>
        }
        level={1}
        onClick={[Function]}
        onClickExpansionToggle={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
      >
        <div
          aria-selected="true"
          className="core-tree-node is-selected"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          role="treeitem"
          tabIndex={-1}
        >
          <div
            className="contents"
            style={
              Object {
                "marginLeft": 20,
              }
            }
          >
            <ProgressRadial
              indeterminate={true}
              size="x-small"
            >
              <div
                className="iui-progress-indicator-radial iui-indeterminate iui-x-small"
              >
                <svg
                  aria-hidden="true"
                  className="iui-radial"
                  viewBox="0 0 32 32"
                >
                  <circle
                    className="iui-track"
                    cx="16"
                    cy="16"
                    r="14"
                  />
                  <circle
                    className="iui-fill"
                    cx="16"
                    cy="16"
                    r="14"
                    style={
                      Object {
                        "strokeDashoffset": 0,
                      }
                    }
                  />
                </svg>
              </div>
            </ProgressRadial>
            <div>
              <Component
                key={[Function]}
              >
                <Component />
              </Component>
            </div>
          </div>
        </div>
      </TreeNode>
    </div>
    <div />
  </div>
</TreeNodeWrapper>
`;

exports[`TreeComponent tests. TreeWithRuleset renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ControlledTreeWrapper
    dataProvider={
      Object {
        "dispose": [Function],
        "getFilteredNodePaths": [Function],
        "getNodeKey": [Function],
        "getNodes": [Function],
        "getNodesCount": [Function],
        "imodel": null,
        "rulesetId": "breakdown-trees-react/RulesList",
      }
    }
    displayGuids={true}
    enableVisibility={false}
    iModel={null}
    loadedRuleset={
      Object {
        "$schema": "../../node_modules/@itwin/presentation-common/Ruleset.schema.json",
        "id": "itwin-breakdown-trees/SpatialBreakdown",
        "rules": Array [
          Object {
            "ruleType": "RootNodes",
            "specifications": Array [
              Object {
                "classes": Array [
                  Object {
                    "arePolymorphic": true,
                    "classNames": Array [
                      "CompositeElement",
                    ],
                    "schemaName": "SpatialComposition",
                  },
                ],
                "groupByClass": false,
                "groupByLabel": false,
                "instanceFilter": "NOT model.IsPrivate AND this.ComposingElement = NULL",
                "relatedInstances": Array [
                  Object {
                    "alias": "model",
                    "isRequired": true,
                    "relationshipPath": Object {
                      "direction": "Backward",
                      "relationship": Object {
                        "className": "ModelContainsElements",
                        "schemaName": "BisCore",
                      },
                      "targetClass": Object {
                        "className": "Model",
                        "schemaName": "BisCore",
                      },
                    },
                  },
                ],
                "specType": "InstanceNodesOfSpecificClasses",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Backward",
                    "relationship": Object {
                      "className": "PhysicalSystemServicesSpatialElements",
                      "schemaName": "BisCore",
                    },
                    "targetClass": Object {
                      "className": "DistributionSystem",
                      "schemaName": "DistributionSystems",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"DistributionSystem\\", \\"DistributionSystems\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "DistributionSystemGroupsDistributionElements",
                      "schemaName": "DistributionSystems",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "Biscore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "CompositeComposesSubComposites",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "CompositeElement",
                      "schemaName": "SpatialComposition",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "CompositeOverlapsSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "SpatialOrganizerReferencesSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "SpatialOrganizerHoldsSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
            "ruleType": "Content",
            "specifications": Array [
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeOverlapsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                      "targetClass": Object {
                        "className": "ISpatialOrganizer",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerHoldsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerReferencesSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "DistributionSystemGroupsDistributionElements",
                        "schemaName": "DistributionSystems",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
            ],
          },
        ],
      }
    }
    optionItems={Array []}
    searchTools={true}
    setIsDisplayGuids={[Function]}
    treeName="TreeWithRulesetTest"
    treeNodeIconMapper={
      TreeNodeFunctionIconInfoMapper {
        "_TreeNodeClassFunctionIconInfoMap": Map {},
        "_functionIconInfos": Array [],
        "_globalFunctionIconInfos": Array [],
        "_groupNodeFunctionIconInfos": Array [],
        "_multipleNodesFunctionIconInfos": Array [],
        "_treeDataProvider": Object {
          "dispose": [Function],
          "getFilteredNodePaths": [Function],
          "getNodeKey": [Function],
          "getNodes": [Function],
          "getNodesCount": [Function],
          "imodel": null,
          "rulesetId": "breakdown-trees-react/RulesList",
        },
      }
    }
  />
</ContextProvider>
`;

exports[`TreeComponent tests. TreeWithRuleset renders correctly with visibility 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ControlledTreeWrapper
    dataProvider={
      Object {
        "dispose": [Function],
        "getFilteredNodePaths": [Function],
        "getNodeKey": [Function],
        "getNodes": [Function],
        "getNodesCount": [Function],
        "imodel": null,
        "rulesetId": "breakdown-trees-react/RulesList",
      }
    }
    displayGuids={true}
    enableVisibility={true}
    iModel={null}
    loadedRuleset={
      Object {
        "$schema": "../../node_modules/@itwin/presentation-common/Ruleset.schema.json",
        "id": "itwin-breakdown-trees/SpatialBreakdown",
        "rules": Array [
          Object {
            "ruleType": "RootNodes",
            "specifications": Array [
              Object {
                "classes": Array [
                  Object {
                    "arePolymorphic": true,
                    "classNames": Array [
                      "CompositeElement",
                    ],
                    "schemaName": "SpatialComposition",
                  },
                ],
                "groupByClass": false,
                "groupByLabel": false,
                "instanceFilter": "NOT model.IsPrivate AND this.ComposingElement = NULL",
                "relatedInstances": Array [
                  Object {
                    "alias": "model",
                    "isRequired": true,
                    "relationshipPath": Object {
                      "direction": "Backward",
                      "relationship": Object {
                        "className": "ModelContainsElements",
                        "schemaName": "BisCore",
                      },
                      "targetClass": Object {
                        "className": "Model",
                        "schemaName": "BisCore",
                      },
                    },
                  },
                ],
                "specType": "InstanceNodesOfSpecificClasses",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Backward",
                    "relationship": Object {
                      "className": "PhysicalSystemServicesSpatialElements",
                      "schemaName": "BisCore",
                    },
                    "targetClass": Object {
                      "className": "DistributionSystem",
                      "schemaName": "DistributionSystems",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"DistributionSystem\\", \\"DistributionSystems\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "DistributionSystemGroupsDistributionElements",
                      "schemaName": "DistributionSystems",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "Biscore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "CompositeComposesSubComposites",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "CompositeElement",
                      "schemaName": "SpatialComposition",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "CompositeOverlapsSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "SpatialOrganizerReferencesSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relationshipPaths": Array [
                  Object {
                    "direction": "Forward",
                    "relationship": Object {
                      "className": "SpatialOrganizerHoldsSpatialElements",
                      "schemaName": "SpatialComposition",
                    },
                    "targetClass": Object {
                      "className": "SpatialElement",
                      "schemaName": "BisCore",
                    },
                  },
                ],
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
            "ruleType": "Content",
            "specifications": Array [
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeOverlapsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                      "targetClass": Object {
                        "className": "ISpatialOrganizer",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerHoldsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerReferencesSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "DistributionSystemGroupsDistributionElements",
                        "schemaName": "DistributionSystems",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
            ],
          },
        ],
      }
    }
    optionItems={Array []}
    searchTools={true}
    setIsDisplayGuids={[Function]}
    treeName="TreeWithRulesetTest"
    treeNodeIconMapper={
      TreeNodeFunctionIconInfoMapper {
        "_TreeNodeClassFunctionIconInfoMap": Map {},
        "_functionIconInfos": Array [],
        "_globalFunctionIconInfos": Array [],
        "_groupNodeFunctionIconInfos": Array [],
        "_multipleNodesFunctionIconInfos": Array [],
        "_treeDataProvider": Object {
          "dispose": [Function],
          "getFilteredNodePaths": [Function],
          "getNodeKey": [Function],
          "getNodes": [Function],
          "getNodesCount": [Function],
          "imodel": null,
          "rulesetId": "breakdown-trees-react/RulesList",
        },
      }
    }
  />
</ContextProvider>
`;
